staging/src/k8s.io/apiserver/pkg/registry/rest/create.go

```diff
@@ -150,16 +152,25 @@ func CheckGeneratedNameError(strategy RESTCreateStrategy, err error, obj runtime
 		return err
 	}
 
-	objectMeta, kind, kerr := objectMetaAndKind(strategy, obj)
+	objectMeta, gvk, kerr := objectMetaAndKind(strategy, obj)
 	if kerr != nil {
 		return kerr
 	}
 
 	if len(objectMeta.GetGenerateName()) == 0 {
+		// If we don't have a generated name, return the original error (AlreadyExists).
+		// When we're here, the user picked a name that is causing a conflict.
 		return err
 	}
 
-	return errors.NewServerTimeoutForKind(kind.GroupKind(), "POST", 0)
+	// If we have a name and generated name, the server picked a name
+	// that already exists.
+	e := errors.NewAlreadyExists(
+		schema.GroupResource{Group: gvk.Group, Resource: gvk.Kind},
+		objectMeta.GetName(),
+	)
+	e.ErrStatus.Message = fmt.Sprintf("%s: the server was not able to generate a valid name for the object", e.ErrStatus.Message)
```

Comment:
Great, this is an improvement already and I could see going with it as is, but:
* We typically want a client to retry. I think it would be best if the message said this.
* If we want auto-retry behavior, we could turn this into a 429. This is just an idea and I'm not sure if I like it yet; maybe @deads2k or @liggitt have a stronger opinion?

We should add a test. A unit test is probably OK, but I think with a webhook it should be possible to deterministically test this in a live cluster. It should be a fun webhook to write, but maybe that's overkill :)

Reply:
> We typically want a client to retry. I think it would be best if the message said this.
Yeah, this is definitely something that it'd make sense for GenerateName, but not for name. Given that 429 is http's TooManyRequests, is this something that can confuse users? In general, I wouldn't expect users relying on generateName to expect an AlreadyExists error, but rather something more specific; should we create a custom error on top of 429 to take advantage of the auto-retry behavior?

> We should add a test. A unit test is probably OK, but I think with a webhook it should be possible to deterministically test this in a live cluster. It should be a fun webhook to write, but maybe that's overkill :)

👍 I'll make sure to update / add new tests once we agree on a solution

Reply:
Ah I keep forgetting, it doesn't need to be 429, we just have to set the Retry-After header to get the retry behavior. So let's keep this status code. If there's an easy way to plumb a response header (in the error? in the context?) let's do that but there might not be, in which case we can save it for a follow up.

Reply:
setting status details RetryAfterSeconds in the error will set this header for you:

https://github.com/kubernetes/kubernetes/blob/release-1.22/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go#L283-L287

Reply:
I pushed some changes and created a new error for generated name conflicts, which also allows to set the RetryAfterSeconds